pghoard/basebackup_delta.py

```diff
@@ -0,0 +1,347 @@
+        snapshot_result.hashes = [
+            SnapshotHash(hexdigest=ssfile.hexdigest, size=ssfile.file_size)
+            for ssfile in snapshot_result.state.files
+            if ssfile.hexdigest
+        ]
+        snapshot_result.files = len(snapshot_result.state.files)
+        snapshot_result.total_size = sum(ssfile.file_size for ssfile in snapshot_result.state.files)
+        snapshot_result.end = now()
+
+        self.log.debug("snapshot result: %r", snapshot_result.json())
+
+        return snapshot_result
+
+    def _upload_hexdigest_from_file(self, file_obj, relative_path, callback_queue):
+        """Upload file into an object with a temporary name, after upload is finished, it will be copied to a new name
+        with a final/real hash"""
+        # length of UUID is 36 + prefix, should not conflict with actual digest file names
+        temp_object_name = f"temp_{uuid.uuid4()}"
+        self._delta_upload_hexdigest(
+            callback_queue=callback_queue,
+            chunk_path=f"{self.compressed_base}/{temp_object_name}",
+            temp_dir=self.temp_base_dir,
+            file_obj=file_obj,
+            relative_path=relative_path
+        )
+
+    def _list_existing_files(self):
+        """Iterate through all manifest files and fetch information about delta files"""
+        all_snapshot_files: Dict[str, SnapshotFile] = {}
+        for backup in self.get_remote_basebackups_info(self.site):
+            if backup["metadata"].get("format") != BaseBackupFormat.delta_v1:
+                continue
+
+            key = os.path.join(self.site_config["prefix"], "basebackup", backup["name"])
+            bmeta_compressed = self.storage.get_contents_to_string(key)[0]
+
+            with rohmufile.file_reader(
+                fileobj=io.BytesIO(bmeta_compressed),
+                metadata=backup["metadata"],
+                key_lookup=lambda key_id: self.site_config["encryption_keys"][key_id]["private"]
+            ) as input_obj:
+                meta = extract_pghoard_delta_v1_metadata(input_obj)
+
+            manifest = meta["manifest"]
+            snapshot_result = manifest["snapshot_result"]
+            backup_state = snapshot_result["state"]
+            files = backup_state["files"]
+            for delta_file in files:
+                snapshot_file = SnapshotFile.parse_obj(delta_file)
+                if snapshot_file.hexdigest:
+                    all_snapshot_files[snapshot_file.hexdigest] = snapshot_file
+
+        return all_snapshot_files
+
+    def _delta_upload_hexdigest(self, *, temp_dir, chunk_path, file_obj, callback_queue, relative_path):
+        start_time = time.monotonic()
+
+        input_size = file_obj.seek(0, os.SEEK_END)
+        file_obj.seek(0)
+
+        uploaded_hash = hashlib.blake2s()
+
+        def update_hash(data):
+            uploaded_hash.update(data)
+
+        with NamedTemporaryFile(dir=temp_dir, prefix=os.path.basename(chunk_path), suffix=".tmp") as raw_output_obj:
```

Comment:
I see your point, that would probably be a nice idea.
My motivation to do like this was to use already existing uploading mechanism (in transfer.py), so that I can re-use all the logic for metrics/retrying/state and no need to introduce my own, which would probably lead to maintenance difficulties.

Reply:
As discussed in private, I'll keep it as is.